engine/src/state.rs

```diff
@@ -184,17 +192,18 @@ impl State {
         self.hasher.record_board_state(&self.board);
     }
 
-    pub fn play_turn(&mut self, piece: Piece, target_position: Position) {
-        let moves = self.board.moves(&self.turn_color);
-
+    pub fn play_turn(&mut self, piece: Piece, target_position: Position) -> Result<(), GameError> {
         // If the piece is already in play, it's a move
         if self.board.piece_already_played(&piece) {
             let current_position = self.board.position(&piece);
             if self.board.pinned(&current_position) {
-                panic!(
-                    "Invalid history: Piece {} at pos {} is pinned!",
-                    piece, current_position
-                );
+                return Err(GameError::InvalidMove {
+                    piece: piece.to_string(),
+                    from: current_position.to_string(),
+                    to: target_position.to_string(),
+                    turn: self.turn,
+                    reason: "Piece is pinned".to_string(),
+                });
```

Comment:
might be worth creating a `InvalidMove::new(piece, from, to, turn, reason)` function for this and the other instances below

Reply:
Can you have functions on enum variants nowadays?